include/envoy/network/io_handle.h

```diff
@@ -7,15 +7,22 @@
 #include "common/common/assert.h"
 
 namespace Envoy {
-namespace Network {
-
-class IoError;
 
 // IoErrorCode::Again is used frequently. Define it to be a distinguishable address to avoid
 // frequent memory allocation of IoError instance.
 // If this is used, IoHandleCallResult has to be instantiated with a deleter that does not
 // deallocate memory for this error.
-#define ENVOY_ERROR_AGAIN reinterpret_cast<IoError*>(0x01)
+#define ENVOY_ERROR_AGAIN reinterpret_cast<Network::IoError*>(alignof(Network::IoError))
```

Comment:
Can I suggest addressing this with a singleton in the impl? This looks like a pretty gnarly thing to put into the interface.

Reply:
That would be nice if feasible, not sure if it is discussed in previous PR.

Reply:
SGTM.

Reply:
Clarifying: this is a pretty gnarly thing anywhere. And it's particularly gnarly IMO to put it in the interface.

Reply:
Yeah, it looks ugly. But IoError is usually wrapped by unique_ptr. In order to use singleton, the unique_ptr has to be declared with a customized deleter type which is pretty much similar like here. And I assume IoSocketHandle is not the only impl that needs EAGAIN to be specialized (at least quic will need it too). So we probably will end up to have different impl's to have same EAGAIN singleton. Then I would argue that we should put this singleton in interface. This will end up to be something similar as what we do with special address currently, except that 'err_' always points to a valid address. As a result, member access like errorCode() and errorDetails() don't need to be wrapped by static methods. 

I'm also not a big fan of singleton, a const static local variable can do the same. But I feel neither of these can a great advantage over  `ENVOY_ERROR_AGAIN`.

Reply:
For the record I don't usually like singletons conceptually in general in all forms (static locals, registries etc). But actually *const* singletons to me make sense. I'd definitely implement it using the static local initializer pattern in the envoy style guide, which cleanly and easily avoids the init-order fiasco.

RE interface vs impl -- a good pattern here is to have a 2-level hierarchy of Interface -> SharedImpl -> platform-specific impl.

Reply:
RE custom-deleter: that's fine. Just have it call a free() method on the class, which allows the derived impl to decide what to do.

I think you might as well use singletons for all the known types, and only allocate instances when its some error not covered in the enum, and the captured and allocated string is needed to propagate a usable error message.

Reply:
> RE custom-deleter: that's fine. Just have it call a free() method on the class, which allows the derived impl to decide what to do.
In interface, only the deleter type needs to be declared which is most like to be `void(*)(IoError*)`. And each impl can be provided their own deleter function. Adding IoError::free() isn't that necessary to me.
> I think you might as well use singletons for all the known types, and only allocate instances when its some error not covered in the enum, and the captured and allocated string is needed to propagate a usable error message.
The enum might expand and errors other than EAGAIN aren't expected to appear often.  Having singletons for each types somewhat defeat the purpose of having enum.

Reply:
RE the deleter -- sounds reasonable. If you can sort out something that isn't too horrifying I'm fine whatever you decide :)

RE singletons for all enums? Sure that's fine; just use a singleton for Again and not for the others if you prefer. In either case the enum is still valuable as it lets client code make semantic decisions about what to do that work across operating systems.